enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,291 @@
+---
+title: operator-status
+authors:
+  - "@awgreene"
+reviewers:
+  - "@ecordell"
+approvers:
+  - "@kevinrizza"
+creation-date: 2020-05-19
+last-updated: yyyy-mm-dd
+status: implementable
+see-also:
+  - "N/A"  
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# operator-status
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+Managing the upgrade of an operator is a core feature of the [Operator Lifecycle Manager (OLM) project](https://github.com/operator-framework/operator-lifecycle-manager). There are instances where an operator is performing a long running critical task during which it should not be interrupted by an upgrade initiated by OLM. OLM must provide operators with the means to communicate whether or not the operator is ready to be upgraded.
+
+## Motivation
+
+Today, OLM does not provide operators with the ability to block upgrades as the operator completes a critical process. Consider the following example:
+
+An operator has three versions available, v1, v2, and v3 where v2 has a long-running process that migrates resources in v1 to the new schema in v2. As of today, if operator v1 was installed with automatic updates enabled, and v2 and v3 became available at the same time, OLM would upgrade the operator from v1 to v2 to v3 without allowing v2's migration process to run. 
+
+The scenario above could place the operator in an unrecoverable state. OLM must provide operators that it manages with a method to commuincate whether or not the operator is ready to be upgraded. 
+
+### Goals
+
+- Provide opreators with the means to communicate whether or not they should be upgraded.
+- Update OLM so operators are not upgraded if they communicate that they should not be upgraded.
+- Preserve existing behavior should the operator not commuincate its upgradeable status.
+
+### Non-Goals
+
+- Support operator version rollback.
+- Provide operators with the means to cancle an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade my operator until I remove the flag.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Operators can communicate update readiness by exposing an `upgradeReadiness` endpoint. This endpoint can communicate readiness by:
+
+- printing 0, indicating that the operator cannot be upgraded at this time.
+- printing 1, indicating that the operator is ready to be upgraded.
+
+This endpoint will be exposed via a [Service resource](https://kubernetes.io/docs/concepts/services-networking/service/#service-resource) that OLM will probe for update readiness prior to upgrading the operator to the next version.
```

Comment:
Say an operator writes `1` to this endpoint, immediately dies somehow, then an upgrade is made available; would OLM not upgrade the operator because it has some other pod status indicating it isn't healthy, or would it attempt to upgrade since it got a positive probe response?

Reply:
As outlined in the enhancement, if the operator is in a ready state and exposes that it is "upgradeable", the upgrade will kickoff. Once an upgrade is kicked off, OLM will not support rollback.